### PR TITLE
Handle IllegalStateException thrown by addShutdownHook

### LINKED ISSUE #932: java.lang.IllegalStateException in related to the changes of class MemoryMXBean
In Java 8 SR 5, the implementation of MemoryMXBeanImpl depends on javax/management/NotificationBroadcasterSupport, which in turns depends on java/util/logging/LogManager.  LogManager will then register a Shutdown hook.
If the JDK has started to shut down when the LogManager tries to register a Shutdown hook, The java.lang.IllegalStateException  will throw.

Standalone test case:
```
public class JMXTest {
    public static void main(String args[]) throws Exception {
        // Add a shutdown hook to artificially slow down the shutdown process
        Thread hookThread = new Thread() {
            public void run() {
                try {
                    System.out.println("The hookThread is about to sleep");
                    Thread.sleep(500);
                    System.out.println("The hookThread awakes");
                } catch (Exception e) {
                }
            }
        };
        Runtime.getRuntime().addShutdownHook(hookThread);
       
        // Initialize the MemoryMXBean on a new Daemon thread
        // (Daemon threads don't prevent shutdown)
        Thread beanThread = new Thread() {
            public void run() {
                System.out.println("The BeanThread is running.");
                MemoryMXBean bean = ManagementFactory.getMemoryMXBean();
                System.out.println("Bean successfully initialized " + bean.getObjectPendingFinalizationCount());
            }

        };
        beanThread.setDaemon(true);
        beanThread.start();

        // Let shutdown happen naturally
    }
}
```

$ ../sdk/jre/bin/java JMXTest
The BeanThread is running.
The hookThread is about to sleep
Bean successfully initialized 0
Exception in thread "MemoryMXBean notification dispatcher" java.lang.IllegalStateException: Shutdown in progress
        at java.lang.ApplicationShutdownHooks.add(ApplicationShutdownHooks.java:77)
        at java.lang.Runtime.addShutdownHook(Runtime.java:222)
        at com.ibm.lang.management.internal.MemoryNotificationThread.lambda$registerShutdownHandler$0(MemoryNotificationThread.java:162)
        at com.ibm.lang.management.internal.MemoryNotificationThread$$Lambda$8.000000000C002D70.run(Unknown Source)
        at java.security.AccessController.doPrivileged(AccessController.java:638)
        at com.ibm.lang.management.internal.MemoryNotificationThread.registerShutdownHandler(MemoryNotificationThread.java:166)
        at com.ibm.lang.management.internal.MemoryNotificationThread.run(MemoryNotificationThread.java:175)
The hookThread awakes

This is the full Java stack from the corresponding system core:

> stack 0x0000000001F9FB00
<1f9fb00>                               Generic special frame
<1f9fb00>       !j9method 0x0000000001F878A8   java/lang/ApplicationShutdownHooks.add(Ljava/lang/Thread;)V
<1f9fb00>       !j9method 0x0000000001F20540   java/lang/Runtime.addShutdownHook(Ljava/lang/Thread;)V
<1f9fb00>       !j9method 0x0000000001FE8BD8   java/util/logging/LogManager.<init>(Ljava/lang/Void;)V
<1f9fb00>       !j9method 0x0000000001FE8BB8   java/util/logging/LogManager.<init>()V
<1f9fb00>       !j9method 0x0000000001FE9260   java/util/logging/LogManager$1.run()Ljava/util/logging/LogManager;
<1f9fb00>       !j9method 0x0000000001FE9280   java/util/logging/LogManager$1.run()Ljava/lang/Object;
<1f9fb00>       !j9method 0x0000000001F2BAF0   java/security/AccessController.doPrivileged(Ljava/security/PrivilegedAction;)Ljava/lang/Object;
<1f9fb00>       !j9method 0x0000000001FE9218   java/util/logging/LogManager.<clinit>()V
<1f9fb00>                               JNI call-in frame
<1f9fb00>                               Generic special frame
<1f9fb00>       !j9method 0x0000000001FE1888   java/util/logging/Logger.demandLogger(Ljava/lang/String;Ljava/lang/String;Ljava/lang/Class;)Ljava/util/logging/Logger;
<1f9fb00>       !j9method 0x0000000001FE18A8   java/util/logging/Logger.getLogger(Ljava/lang/String;)Ljava/util/logging/Logger;
<1f9fb00>       !j9method 0x0000000001FD7DA8   com/sun/jmx/remote/util/ClassLogger.<init>(Ljava/lang/String;Ljava/lang/String;)V
<1f9fb00>       !j9method 0x0000000001FD6720   javax/management/NotificationBroadcasterSupport.<clinit>()V
<1f9fb00>                               JNI call-in frame
<1f9fb00>                               Generic special frame
<1f9fb00>       !j9method 0x0000000001FD3BA0   com/ibm/java/lang/management/internal/MemoryMXBeanImpl.<init>()V
<1f9fb00>       !j9method 0x0000000001FD44A0   com/ibm/java/lang/management/internal/MemoryMXBeanImpl.<clinit>()V
<1f9fb00>                               JNI call-in frame
<1f9fb00>                               Generic special frame
<1f9fb00>       !j9method 0x0000000001FD2E58   java/lang/management/ManagementFactory.getMemoryMXBean()Ljava/lang/management/MemoryMXBean;
<1f9fb00>       !j9method 0x0000000001FC93B0   JMXTest$2.run()V

### PR DESCRIPTION
**Pull request by keithc-ca**
It is possible that the VM is in the process of shutting down: we shouldn't start the notification loop in that case because it would never terminate.

Fixes #932.

### CODE REVIEW

**Comment by pshipton** [comment]
Can we add the test case that covers this case to the test suite?

**Comment by keithc-ca** [comment]
I will add (an adjusted) test case to the test suite.

**Comment by keithc-ca** [comment]
@llxia has had a preliminary look over the test changes.

**Comment by pshipton** [comment]
@llxia @smlambert any concerns with the test changes?

**Comment by llxia** [comment]
Test changes look good to me.

File: test/Java8andUp/src/org/openj9/test/java/lang/management/MemoryMXBeanShutdownNotification.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package org.openj9.test.java.lang.management;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+
+import org.testng.Assert;
+import org.testng.AssertJUnit;
```
Comment (smlambert) [inline test/Java8andUp/src/org/openj9/test/java/lang/management/MemoryMXBeanShutdownNotification.java]: You don't need or use this, as this is a newly written test (not a converted JUnit test), which reminds that we should update the TestExample to use the Assert class directly, as you have done.

Comment (keithc-ca) [reply test/Java8andUp/src/org/openj9/test/java/lang/management/MemoryMXBeanShutdownNotification.java]: https://github.com/eclipse/openj9/pull/956 removes that unused import.

**Pull request merged**